### PR TITLE
Document legacy build vs BuildKit differences

### LINKED ISSUE #12231: Recommend Buildkit for using multi-stage builds
File: [develop/develop-images/multistage-build.md](https://docs.docker.com/develop/develop-images/multistage-build/)

When people started off using Docker a couple of versions ago, their default is set to not use Buildkit. However, Buildkit is required to properly execute `-t release` stages. Without Buildkit, other stages of the same Dockerfile are executed as well although they shouldn't.

Please add a hint that the described features are only properly usable with Buildkit enabled.

### PR DESCRIPTION
**Pull request by andreisilviudragnea**
### Proposed changes

I documented how legacy engine executes all stages of a multi-stage Dockerfile from the beginning of the Dockerfile up to the target stage, even if target stage does not depend on some intermediate stages. When BuildKit is enabled, this (possible bug) does not happen and only transitive dependency stages of the target stage get executed.

I also documented that BuildKit is not enabled by default on all platforms (it is on macOS, but not on Linux) and that multi-stage builds should always be executed with BuildKit enabled in order to avoid the problem described above.

### Related issues

Fixes #12231

### CODE REVIEW

**Comment by andreisilviudragnea** [comment]
@craig-osterhout When will this be reviewed and merged?

**Comment by usha-mandya** [comment]
@andreisilviudragnea Could you rebase the PR?

**Comment by andreisilviudragnea** [comment]
@usha-mandya Done.

**Comment by usha-mandya** [comment]
Thank you @andreisilviudragnea. One of our team members will take a look and get back to you.

File: build/building/multi-stage.md [hunk 1]
```diff
@@ -205,3 +205,83 @@ RUN g++ -o /binary source.cpp
...
+FROM base AS stage2
+RUN echo "stage2"
+```
+
+- Output for BuildKit enabled:
+
+```console
+$ DOCKER_BUILDKIT=1 docker build --no-cache -f Dockerfile --target stage2 .
+```
+```console
```
Comment (dvdksn) [inline build/building/multi-stage.md:234]:
```suggestion
```

File: build/building/multi-stage.md [hunk 2]
```diff
@@ -205,3 +205,83 @@ RUN g++ -o /binary source.cpp
...
+ => [base 2/2] RUN echo "base"                                                                                                                                                                             0.1s
+ => [stage2 1/1] RUN echo "stage2"                                                                                                                                                                         0.2s
+ => exporting to image                                                                                                                                                                                     0.0s
+ => => exporting layers                                                                                                                                                                                    0.0s
+ => => writing image sha256:f55003b607cef37614f607f0728e6fd4d113a4bf7ef12210da338c716f2cfd15                                                                                                               0.0s
+```
+
+- Output for BuildKit disabled:
```
Comment (dvdksn) [inline build/building/multi-stage.md:249]:
```suggestion
On the other hand, building the same target without BuildKit results in all stages being processed:
```

File: build/building/multi-stage.md [hunk 3]
```diff
@@ -205,3 +205,83 @@ RUN g++ -o /binary source.cpp
...
+
+```console
+$ DOCKER_BUILDKIT=0 docker build --no-cache -f Dockerfile --target stage2 .
+```
+```console
```
Comment (dvdksn) [inline build/building/multi-stage.md:254]:
```suggestion
```

File: build/building/multi-stage.md [hunk 4]
```diff
@@ -205,3 +205,83 @@ RUN g++ -o /binary source.cpp
...
+stage2
+Removing intermediate container bbc025b93175
+ ---> 09fc3770a9c4
+Successfully built 09fc3770a9c4
+```
+
+`stage1` gets executed when BuildKit is disabled, even if `stage2` does not depend on it.
+
+BuildKit is not enabled by default on all platforms. For example, while it is enabled on macOS, it is
+NOT on Linux.
```
Comment (dvdksn) [inline build/building/multi-stage.md:283]: 1) on what platforms is it enabled by default? (how about Windows?)
2) how can I check if I am using BuildKit by default?

File: build/building/multi-stage.md [hunk 5]
```diff
@@ -205,3 +205,83 @@ RUN g++ -o /binary source.cpp
...
+
+Docker multi-stage builds should always be run with
+[BuildKit enabled](/develop/develop-images/build_enhancements/)
+to avoid the problem described above.
```
Comment (dvdksn) [inline build/building/multi-stage.md:287]:
```suggestion
Always run multi-stage builds with
[BuildKit enabled](/develop/develop-images/build_enhancements/)
for better performance.
```

**Review by dvdksn — CHANGES_REQUESTED** [review]
I think saying "moby vs buildkit" is confusing, in part since buildkit is also a moby project 😅 I think I would prefer if we could refer to it as "legacy vs buildkit", as we do on [this page](https://docs.docker.com/develop/develop-images/build_enhancements/)

File: build/building/multi-stage.md (hunk 4 above)
Comment (jedevc) [reply build/building/multi-stage.md:283]: 1. I think it's enabled by default on Docker Desktop, but not on our Linux packages - the default should be switching to BuildKit everywhere (with buildx :tada:) in the upcoming release of 22.06, see https://github.com/docker/cli/pull/3676.
2. I don't actually know :scream_cat: I know you can look at the configuration file for your installation, e.g. `/etc/docker/daemon.json`, and see if `buildkit` is enabled under the `features` key.

Comment (dvdksn) [reply build/building/multi-stage.md:283]: OK, gotcha! Let me suggest another edit based on this info.

Comment (dvdksn) [inline build/building/multi-stage.md:283]: Given that we are about to make BuildKit the default in the near future, we can simplify to this:

```suggestion
BuildKit is enabled by default if you use Docker Desktop.
```

File: build/building/multi-stage.md (hunk 5 above)
Comment (andreisilviudragnea) [reply build/building/multi-stage.md:287]: While I agree with the better performance part, it also changes the semantics of the build.

This bug showed up in our build when enabling BuildKit on a Linux installation. Do you think this should be logged as a bug too? Is it a bug or a feature?

**Review by dvdksn — APPROVED** [review]
lgtm, thanks a lot @andreisilviudragnea

**Pull request merged**